### PR TITLE
feat(firebase_auth): add custom auth domain setter to Firebase Auth

### PR DESCRIPTION
**Pull request by Andrey-K-Calven**
## Description

https://github.com/firebase/flutterfire/pull/11925 allowed, previously web exclusive, authDomain property to be used on mobile platform; however, actually using it has significant limitations:
  1. It requires creating a separate named FirebaseApp with authDomain property on top of the `[DEFAULT]` `FirebaseApp` generated from `GoogleService-Info.plist`/`google-services.json` because there's no way to set authDomain via those files.
  2. Changing authDomain requires deleting and recreating the FirebaseApp.
  
This change addresses those by adding a custom auth domain setter to `FirebaseAuth`, plugging into the existing corresponding setters in the native implementations of FirebaseAuth. With this setter authDomain can be set and reset at will on the `[DEFAULT]` `FirebaseApp`'s `FirebaseAuth`. This allows for custom auth domains to be used with the go-to style of `FirebaseApp` initialisation on mobile platforms and is consistent with the way this is handled in the native implementations.

## Related Issues

expands on https://github.com/firebase/flutterfire/pull/11925
closes https://github.com/firebase/flutterfire/issues/12517

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by russellwheatley** [comment]
Hey @Andrey-K-Calven - thanks for the PR! We're discussing internally on how to move this forward, will let you know what results from the discussions 👍

**Comment by Andrey-K-Calven** [comment]
> Hey @Andrey-K-Calven - thanks for the PR! We're discussing internally on how to move this forward, will let you know what results from the discussions 👍

Hi, @russellwheatley. Thanks for looking at this, did you have a chance to have that discussion? We'd like to use this approach for our auth domain needs.

**Comment by Lyokone** [comment]
Hey @Andrey-K-Calven, we finally had the green light for this API. Do you mind merging with master to be sure that tests are passing with all the latest changes?

File: packages/firebase_auth/firebase_auth/windows/messages.g.cpp [hunk 1]
```diff
@@ -2752,14 +2777,10 @@ void FirebaseAuthHostApi::SetUp(flutter::BinaryMessenger* binary_messenger,
...
                   std::get<std::string>(encodable_email_arg);
               const auto& encodable_action_code_settings_arg = args.at(2);
-              const PigeonActionCodeSettings* action_code_settings_arg =
-                  nullptr;
-              if (!encodable_action_code_settings_arg.IsNull()) {
-                action_code_settings_arg =
-                    &(std::any_cast<const PigeonActionCodeSettings&>(
-                        std::get<CustomEncodableValue>(
-                            encodable_action_code_settings_arg)));
-              }
```
Comment (Lyokone) [inline packages/firebase_auth/firebase_auth/windows/messages.g.cpp:2762]: I think this should be put back, not related to the modifications

**Review by Lyokone — CHANGES_REQUESTED** [review]

File: packages/firebase_auth/firebase_auth/test/firebase_auth_test.dart [hunk 2]
```diff
@@ -230,6 +231,37 @@ void main() {
...
+                projectId: 'projectId'));
+
+        FirebaseAuthPlatform.instance =
+            FakeFirebaseAuthPlatform(customAuthDomain: 'foo');
+        auth = FirebaseAuth.instanceFor(app: app);
+
+        expect(auth.customAuthDomain, 'foo');
+        if (defaultTargetPlatform == TargetPlatform.windows || kIsWeb) {
+          try {
+            auth.customAuthDomain = 'bar';
```
Comment (TarekkMA) [inline packages/firebase_auth/firebase_auth/test/firebase_auth_test.dart:252]: Why not testing the getter as well for windows

File: packages/firebase_auth/firebase_auth/windows/messages.g.cpp (hunk 1 above)
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/windows/messages.g.cpp:2762]: Another manual update. We should probably make a script for these.

File: packages/firebase_auth/firebase_auth/test/firebase_auth_test.dart (hunk 2 above)
Comment (russellwheatley) [reply packages/firebase_auth/firebase_auth/test/firebase_auth_test.dart:252]: As mentioned in chat, this is just to illustrate what happens for each platform. The getter will always be null as we throw exception on setter.

**Review by TarekkMA — APPROVED** [review]

**Review by Lyokone — APPROVED** [review]

**Pull request merged**